js/buffer.ts

```diff
@@ -0,0 +1,219 @@
+   * portion of the buffer
+   */
+  get length() {
+    return this.buf.byteLength - this.off;
+  }
+
+  /** Cap returns the capacity of the buffer's underlying byte slice, that is,
+   * the total space allocated for the buffer's data.
+   */
+  cap(): number {
+    return this.buf.buffer.byteLength;
+  }
+
+  /** truncate() discards all but the first n unread bytes from the buffer but
+   * continues to use the same allocated storage.  It throws if n is negative or
+   * greater than the length of the buffer.
+   */
+  truncate(n: number): void {
+    if (n === 0) {
+      this.reset();
+      return;
+    }
+    if (n < 0 || n > this.length) {
+      throw Error("bytes.Buffer: truncation out of range");
+    }
+    this._reslice(this.off + n);
+  }
+
+  /** reset() resets the buffer to be empty, but it retains the underlying
+   * storage for use by future writes. reset() is the same as truncate(0)
+   */
+  reset() {
+    this._reslice(0);
+    this.off = 0;
+  }
+
+  /** _tryGrowByReslice() is a inlineable version of grow for the fast-case
+   * where the internal buffer only needs to be resliced. It returns the index
+   * where bytes should be written and whether it succeeded.
+   */
+  private _tryGrowByReslice(n: number): [number, boolean] {
+    const l = this.buf.byteLength;
+    if (n <= this.cap() - l) {
+      this._reslice(l + n);
+      return [l, true];
+    }
+    return [0, false];
+  }
+
+  private _reslice(len: number): void {
+    assert(len <= this.buf.buffer.byteLength);
+    this.buf = new Uint8Array(this.buf.buffer, 0, len);
+  }
+
+  /** read() reads the next len(p) bytes from the buffer or until the buffer
+   * is drained. The return value n is the number of bytes read. If the
+   * buffer has no data to return, eof in the response will be true.
+   */
+  async read(p: ArrayBufferView): Promise<ReadResult> {
+    if (this.empty()) {
+      // Buffer is empty, reset to recover space.
+      this.reset();
+      if (p.byteLength === 0) {
+        return { nread: 0, eof: false };
+      }
+      return { nread: 0, eof: true };
+    }
+    const nread = copyBytes(p as TypedArray, this.buf.subarray(this.off));
+    this.off += nread;
+    return { nread, eof: false };
+  }
+
+  async write(p: ArrayBufferView): Promise<number> {
+    const m = this._grow(p.byteLength);
+    return copyBytes(this.buf, p as TypedArray, m);
+  }
+
+  /** _grow() grows the buffer to guarantee space for n more bytes.
+   * It returns the index where bytes should be written.
+   * If the buffer can't grow it will throw with ErrTooLarge.
+   */
+  private _grow(n: number): number {
+    const m = this.length;
+    // If buffer is empty, reset to recover space.
+    if (m === 0 && this.off !== 0) {
+      this.reset();
+    }
+    // Fast: Try to grow by means of a reslice.
+    const [i, ok] = this._tryGrowByReslice(n);
+    if (ok) {
+      return i;
+    }
+    const c = this.cap();
+    if (n <= c / 2 - m) {
```

Comment:
Note that, unlike rust, javascript does not have the concept of an "integer division".
Maybe you want to coerce it to an int `(c / 2) | 0` or `(c >> 1)` so v8 doesn't have to convert c, n and m to doubles.